### ISSUE TITLE
Restoring from different host saying `wrong password or no key found`

### ISSUE THREAD

**Issue by emmetog** [restic/restic#923]
Recently I backed up my entire home folder (around 75Gb), tested that the restore worked (it did) and then reinstalled my entire computer, wiping everything thinking I had everything safely backed up. Bad move. Now I keep getting the error `unable to open repo: wrong password or no key found` when I try to restore on the same (now wiped) laptop :sob: 

Question: are the password keys somehow linked to a host or is it probable that my repo is corrupt?

## Output of `restic version`
```
debug log file /tmp/restic.log
debug enabled
restic 0.5.0 (v0.5.0-124-gbcf97eb)
compiled with go1.7.4 on linux/amd64
```

As you can see I've enabled the debug mode to see the logs.

## How did you start restic exactly? (Include the complete command line)
```
$ DEBUG_LOG=/tmp/restic.log AWS_ACCESS_KEY_ID=XXXXXX AWS_SECRET_ACCESS_KEY="XXXXXX" restic --repo s3:s3.amazonaws.com/XXXXXX check
debug log file /tmp/restic.log
debug enabled
enter password for repository: 
unable to open repo: wrong password or no key found
```

Here's an extract from the `tmp/restic.log` log:
```
2017/04/19 00:14:23 restic/main.go:48	main.main	1	main []string{"restic", "--repo", "s3:s3.amazonaws.com/XXXXXX", "check"}
2017/04/19 00:14:23 restic/global.go:374	main.open	1	parsing location s3:s3.amazonaws.com/XXXXXX
2017/04/19 00:14:23 restic/global.go:356	main.parseConfig	1	opening s3 repository at s3.Config{Endpoint:"s3.amazonaws.com", UseHTTP:false, KeyID:"XXXXXX", Secret:"XXXXXX", Bucket:"XXXXXX", Prefix:"restic"}
2017/04/19 00:14:23 s3/s3.go:36	s3.Open	1	open, config s3.Config{Endpoint:"s3.amazonaws.com", UseHTTP:false, KeyID:"XXXXXX", Secret:"XXXXXX", Bucket:"XXXXXX", Prefix:"restic"}
2017/04/19 00:14:27 s3/s3.go:291	s3.(*s3).List	1	listing key
2017/04/19 00:14:27 restic/lock.go:112	main.unlockAll	1	unlocking 0 locks
```

## What backend/server/service did you use?
S3

## Expected behavior
Everything should work with the same password from a different host

## Actual behavior
I cannot connect to the repo again, the error is `unable to open repo: wrong password or no key found`.

## Steps to reproduce the behavior
See below

## Do you have any idea what may have caused this?
Does restic use the hostname in the calculations of the keys? Is a key bound only to one single host? Or is it more probable that my repo is corrupt?

If it's corrupt then the only thing that I can think of that could have corrupted it is that I used `--force`. Could that be the culprit?

Here's what I did:

* I backed up entire home dir to S3 (75Gb, it took about 3 hours)
* I made a few changes to some files and tried backing up again, but because the repo was so big it was taking +30 hours even though only very few files had changed. I read in another issue in this project that `--force` saved the bandwidth of getting the metadata from S3 so I did that and it was much faster, completing in less than an hour
* I tested a restore to a different dir although I didn't let it complete, all seemed to work for the few directories I saw being restored (it looked like it was going to take aaaages though, but that's ok)
* At this point I wiped my laptop
* Installed restic on new laptop and tried to restore but running into this

 Any insight would be much appreciated!

**Comment by fd0** [comment]
Thanks for the comprehensive report. I'm very sorry for the bad user experience! There's nothing in the repo or in the data structures that is bound to a specific host/installation, you should be able to restore to a different host (provided you have the right password of course). Also I'm sure that using `--force` did not cause this.

We recently made structural changes in the s3 backend (among others), which caused a very similar bug (see #917, solved by #918). I'm positive that this also is just a failure to access the key file stored in the repo, and no data has been lost.

As a first step, could you please try the released version 0.5.0? If that doesn't you can also checkout the commit one step before structure change, by executing the following commands:
```
$ git reset --hard e73038cbcb558785d277b1d6ab238747b7fa5f96
```
Then compile restic as usual and please report back!

**Comment by fd0** [comment]
Further ideas:
 * Are you sure that the repository path (bucket and path within the bucket) is correct? Maybe you used a sbudir within the bucket? Check with the s3 console what's in your bucket, search for a directory that contains the file `config`, that's your repository
 * Try using the same version of restic as the one you used to create the repo

**Comment by emmetog** [comment]
Resolved! The issue was the subdirectory in the bucket as you suggested. It looks like when the backup was performed the first time restic created a `s3.amazonaws.com` subdirectory and put the actual repository inside that. As far as I remember this happened behind the scenes during the backup, I didn't specify that, just the name of the bucket. When I performed the original backup I think I was using a self built binary of the latest git revision (as of 2 days ago).

However when it comes to other operations like snapshots, restore, keys, etc, it looks like restic needs that subdirectory to be explicitly set. I've checked on the released version 0.5.0 and on the latest git revision with the same results.

In any case my problem is resolved by specifying the subdirectory when doing the restore, like this:
```
restic --repo s3:s3.amazonaws.com/bucket-name/s3.amazonaws.com restore ...
```

A huge thanks to you @fd0 for your support!

**Comment by fd0** [comment]
Heh, glad you found the cause for this issue! Not being able to restore a backup is very stressful... I'll add a check for the presence of the config file and an appropriate error message, this is tracked in #925.

restic does not have any code which adds such a subdir, so I suspect that you may have either typed the hostname twice (in the repo specification string), or had an additional character `/` at the beginning of the host name? Could you maybe check with the (now restored) shell history files? That'd be interesting to know, so we can improve the user interface.
